### ISSUE TITLE
java.lang.StackOverflowError during "Requesting Java AST from selection"

### ISSUE THREAD

**Issue by nlisker** [eclipse-jdt/eclipse.jdt.core#3273]
Eclipse soft crashes during hover over some fields with a SOE and asks to restart. This also breaks the compilation, as after this error, many red lines appear saying that various methods can't be found suddenly.
![image](https://github.com/user-attachments/assets/264f5dbb-98a5-466f-be5d-089429ad053c)
![image](https://github.com/user-attachments/assets/c82af755-eae9-463c-8c62-4beb8b42087b)

The first lines of the stack traces are different every time, but the recursion is always the same. Here is the log with various cases of these errors: [crashes.log](https://github.com/user-attachments/files/17687075/crashes.log)

Here is the infinite recursion (I don't know what the entry point was):
```
	at org.eclipse.jdt.internal.compiler.Compiler.accept(Compiler.java:365)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.accept(CompilationUnitResolver.java:226)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.accept(CompilationUnitResolver.java:221)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.askForType(LookupEnvironment.java:391)
	at org.eclipse.jdt.internal.compiler.lookup.PackageBinding.getTypeOrPackage(PackageBinding.java:275)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.findImport(CompilationUnitScope.java:643)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.resolveImports(CompilationUnitScope.java:325)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.checkAndSetImports(CompilationUnitScope.java:284)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment$CompleteTypeBindingsSteps.perform(LookupEnvironment.java:191)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.completeTypeBindings(LookupEnvironment.java:602)
	at org.eclipse.jdt.internal.compiler.Compiler.accept(Compiler.java:365)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.accept(CompilationUnitResolver.java:226)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.accept(CompilationUnitResolver.java:221)
```

While I can reproduce this 100% of the times, unfortunately I didn't manage to create a minimal example despite many hours of chopping off parts of the big project where this occurs. This is because the error becomes harder and harder to reproduce the more code is removed, which means that there is a certain load affecting this issue.
I can't post a whole project here, but if there is a way to send the project or post it somewhere private I will do so.

Tested on
Eclipse 4.33 and also build id: I20241101-1800
Java 22
Win 10

**Comment by srikanth-sankaran** [comment]
> I can't post a whole project here, but if there is a way to send the project or post it somewhere private I will do so.

I don't have a place to suggest, but if you organize such a place, I offer to investigate. My ideal reproducer will be a plain vanilla SDK project without layers like maven, gradle etc if possible - I am not familiar with these and it will impede progress. Thanks

**Comment by nlisker** [comment]
I sent you an invite to the private repo with instructions in the only open issue there. Thank you for taking a look!

Amber features take priority :)

**Comment by srikanth-sankaran** [comment]
See also 

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1081 
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1490

the former has some reduced test case.

**Comment by srikanth-sankaran** [comment]
In the meantime you may want to check if the hover that triggers the crash is over some symbol statically imported given static imports feature prominently in the other tickets with SOE.

**Comment by nlisker** [comment]
I'll try to narrow it down. For now I'm even getting multiple occurrences
![image](https://github.com/user-attachments/assets/5e14e903-95e1-4693-af46-c6bad7c93f8c)

**Comment by nlisker** [comment]
Also, the crash can also happen on clicking a tab in the editor (switching files), not just on hover.

**Comment by srikanth-sankaran** [comment]
This looks nasty. Hopefully we can get to the bottom of it soon. 

Deeper stack traces may help. 

Also at some point we can consider a version of eclipse that would log some data. We need to identify _what_ to log.

**Comment by nlisker** [comment]
I think managed to narrow it down to a Web Tools feature that is installed in the background. When an Eclipse with that component opens the project it somehow corrupts it, then any Eclipse that will open it should see the same issue. I'll update the reproduction instructions. Even if it's a bug with a plugin, it bubbles to a full JDT stack trace as shown in the first comment (which is why it's so hard to find where the problem is).

**Comment by nlisker** [comment]
Instructions updated with screenshots.

**Comment by srikanth-sankaran** [comment]
Not yet. I'll spend time on it tomorrow.

**Comment by srikanth-sankaran** [comment]
Drowned in this https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3310 🙂 but I can context switch tomorrow

**Comment by srikanth-sankaran** [comment]
There are some recent, completely unrelated changes to JDT that are blocking me from setting up a development/debugging environment - obviously it is not enough to reproduce, I need to be able to debug the reproduced set up. I need to work with a team mate to get unblocked but he is unavailable today. Will check with him tomorrow. This has bubbled to near the top of priority list, so hopefully will be resolved soon. Stay tuned. Thanks for your patience

**Comment by srikanth-sankaran** [comment]
(The issue referred https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3190#issuecomment-2500203575)

**Comment by srikanth-sankaran** [comment]
![Screenshot 2024-11-28 152521](https://github.com/user-attachments/assets/82f78661-923f-477b-b69d-0c09b731a697)

After installing WST and adding JBoss server, Eclipse wants to restart and upon restart I get this dialog.

Is this expected or materially affects things ?? Sorry a bit clueless here

**Comment by nlisker** [comment]
Never seen this. I use JDK 22/23. The entries there say "requires JavaSE version 21" but you are using 20.

**Comment by srikanth-sankaran** [comment]
> Never seen this. I use JDK 22/23. The entries there say "requires JavaSE version 21" but you are using 20.

I have no idea where that JDK20 reference is coming from, but that doesn't seem to matter, 

Despite that, fairly quickly I am able to reproduce the SOE, so the instructions are good. 

Investigating ...

**Comment by srikanth-sankaran** [comment]
This renders Eclipse unusable pretty quickly in the problem scenario :-(

**Comment by nlisker** [comment]
Thanks, added Stephan.

I still have a few confusing points here.
* If the errors come from these static imports, why could I not reproduce it only on the `shared` module? The error happens on a different module, and even more so only on that specific one I uploaded. The project has more modules that import the `shared` one, but they don't exhibit this error.
* Why could we not reproduce this on a "plain" Eclipse and needed the WST+servers setup to cause this?
* Why, after the error happens once, opening it in a "plain" Eclipse is suddenly reproducible? It looks like something got corrupted.
* I couldn't manage to create a small reproducible. The more I removed the less frequent the error became even if the offending classes were there. Is it triggered when a certain amount of data is processed?

**Comment by nlisker** [comment]
> This renders Eclipse unusable pretty quickly in the problem scenario :-(

Yes, after many hours of figuring out how to reliably and most-minimally reproduce it, I stopped working on the offending module and never came back :) Hopefully I will be able to go back to that part of the code soon.

BTW, is there a workaround? Imports are not strictly necessary I believe, so if I can temporarily fully qualify the classes and get rid of the problematic imports I will be happy to do so until this is resolved.

**Comment by srikanth-sankaran** [comment]
Let us wait for @stephan-herrmann to analyze this - I know very little about JPMS and ECJ's implementation having pretended for a long while Java 9 didn't happen at all and we hop skip jumped from Java 8 to 10 :) But if I were to venture a guess: Could it be that the Split package situation arises only when WST server is installed ? (I will admit I don't even know if this question makes sense.)

**Comment by stephan-herrmann** [comment]
> added Stephan

reproduced also here :)

**Comment by stephan-herrmann** [comment]
> In the latter case, resolving the import `import static com.sanctorum.game.shared.entities.Minion.Characteristics.*;` causes another fresh `CompilationUnitDeclaration` for `Minion.java` to be added to `org.eclipse.jdt.internal.compiler.Compiler.unitsToProcess` notwithstanding the fact that the import is originating from the self same CUD.
> 
> This somehow seems to be connected to the package `com.sanctorum.game.shared.entities` being a `SplitPackageBinding` (from <unnamed>, com.sanctorum.game.shared)

Yep, this is quite close to the root bug, I believe. I can see that we are finding the type in question both as a SourceTypeBinding associated with a proper SourceModuleBinding **and** as an SourceType which is reported to belong to the unnamed module. This latter association is wrong, because the enclosing project is modular.

This happens because the resolved classpath entry towards the project has only these extra attributes: `[attributes:gradle_scope=main,gradle_used_by_scope=main,test]`, where it should in fact also have `module=true`.

The entry is resolved from the following container:
```
        <classpathentry kind="con" path="org.eclipse.buildship.core.gradleclasspathcontainer">
                <attributes>
                        <attribute name="org.eclipse.jst.component.dependency" value="/WEB-INF/lib"/>
                </attributes>
        </classpathentry>
```

Two questions:
* does the gradleclasspathcontainer properly set `module=true` on resolved entries, normally?
* if so, does any wst stuff interfere with this container resolution?
Once the resolved container is persisted this would explain that the bug sticks even when the set of plug-ins that trigger the bug is no longer used.

Possible directions of action:
* ensure that JDT recognizes the schizophrenia of **the same type** appearing with different module associations, *while remaining capable of detecting truly duplicate types from different modules*. Wish me luck!
* chase down the source of the incomplete resolved container entry.

Stay tuned.

**Comment by nlisker** [comment]
I think I can already answer "does the gradleclasspathcontainer properly set module=true on resolved entries, normally?".

The container itself is always non-modular, but the contained artifacts might or might not be. Here is a typical scenario:
![image](https://github.com/user-attachments/assets/771c8a74-ee25-4eef-bbbf-367020f73869)

This is because tests might want to run in a non-modular setup. It is actually a downstream configuration from Eclipse only allowing 1 module-info per project and not per source. I can probably re-find the discussions about the setup of this container on the Gradle/Buildship GitHub.

**Comment by nlisker** [comment]
Also might be of interest:
```
                <attributes>
                        <attribute name="org.eclipse.jst.component.dependency" value="/WEB-INF/lib"/>
                </attributes>
```
JST (Java Server Tools) builds on top of WST (Web Server Tools). When the Java servers are added, it installs JST. This might answer why I found a connection to the WST components that are installed in the background.

**Comment by stephan-herrmann** [comment]
Next find:

The CancelableNameEnvironment used by ASTParser implicitly adds ALL-UNNAMED to the required modules, based on `!excludeTestCode` and presence of a "test" source folder. This is where the unnamed module initially enters the picture.

**Comment by stephan-herrmann** [comment]
> Yep, this is quite close to the root bug, I believe. I can see that we are finding the type in question both as a SourceTypeBinding associated with a proper SourceModuleBinding **and** as an SourceType which is reported to belong to the unnamed module. This latter association is wrong, because the enclosing project is modular.

The final (?) fix in #3373 actually goes the opposite direction: Given that the focus project is not modular, and given that the shared project is on the classpath, not modulepath, then **no** module (other than JRE) should be relevant.

**Comment by nlisker** [comment]
Thank you Stephan. When the fix will be available in the nightly builds I will test it.

**Comment by stephan-herrmann** [comment]
For posterity some attempts at answering:

> I still have a few confusing points here.
> 
> * If the errors come from these static imports, why could I not reproduce it only on the `shared` module? The error happens on a different module, and even more so only on that specific one I uploaded. The project has more modules that import the `shared` one, but they don't exhibit this error.

Static imports pointing into their own compilation unit is one ingredient.

Another ingredient is that different parts of the implementation had different views on the same class and package: To resolve the import, the class was searched in the unnamed module (because the client project is not modular, and the classpath entry pointing to the shared project didn't specify `module=true`). When the class was found we noticed that the containing package fragment root (here: source folder) contains `module-info.java` so we associate the class with that module. But when the recursive lookup tries to resolve its import it still doesn't find its own class via the unnamed module.

This is the schizophrenia I mentioned elsewhere: the implementation "thought" that those are two different packages and classes, one set in the unnamed module the other set in a named module, where in fact both sets are the same.
 
> * Why could we not reproduce this on a "plain" Eclipse and needed the WST+servers setup to cause this?

It is possible that WST+servers influence resolution of the gradle classpath container.  

Which influence exactly this is, I did not investigate, since the error is clearly in JDT.

For completeness: other required ingredients for the bug to occur are: the focus project must have a source folder marked as "test". The shared project must have some non-modular dependency.

> * Why, after the error happens once, opening it in a "plain" Eclipse is suddenly reproducible? It looks like something got corrupted.

The resolved entries (what you seen when you drill into the container in the package explorer) are persisted. So until the container needs resolving again this persistent state will also determine behavior in a WST-less workbench.
 
> * I couldn't manage to create a small reproducible. The more I removed the less frequent the error became even if the offending classes were there. Is it triggered when a certain amount of data is processed?

The number of necessary factors (see above) for reproducing illustrates why extracting a small reproducer was near impossible without debugging.

**Comment by nlisker** [comment]
Thanks for the detailed answers. One question I have remaining is about

> The resolved entries (what you seen when you drill into the container in the package explorer) are persisted.

Where are they persisted? I'm wondering what I would need to clean/recompute in such a case.

**Comment by stephan-herrmann** [comment]
> Thanks for the detailed answers. One question I have remaining is about
> 
> > The resolved entries (what you seen when you drill into the container in the package explorer) are persisted.
> 
> Where are they persisted? I'm wondering what I would need to clean/recompute in such a case.

I believe this is in file `WORKSPACE/.metadata/.plugins/org.eclipse.core.resources/.projects/YOUR_PROJECT/org.eclipse.jdt.core/state.dat`. I have never tried removing such a file, a clean build should be the safer option to achieve the same ... have you tried that?

**Comment by nlisker** [comment]
I've updated to I20241201-1800 and at least for 5 minutes I don't get a crash, which is a record time, so looks fixed! Took me a bit to realize that the nightly builds quietly switched from the I-34 to the I-35 update repository last month.

> a clean build should be the safer option to achieve the same ... have you tried that?

In the following scenario:
Open with plain Eclipse -> no crash
Open with servers Eclipse -> crash
Open with plain Eclipse -> crash
A clean-rebuild on the plain Eclipse did not help. Closing the project and re-opening did not help either. So did removing the project only from the workspace and re-importing. This was part of the confusion on where the data is stored. What did work was removing from the workspace + from the disk and re-cloning from the remote repo.

By the way, a reminder to check if the issues from https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3273#issuecomment-2471791143 were also solved in case you didn't.
